### PR TITLE
Added gallery example 51 on usage of OSM coastlines in GMT

### LINKED ISSUE #3041: Add an example for the usage of OpenStreetMap coastlines to the gallery/CookBook
[I was asked](https://forum.generic-mapping-tools.org/t/accuracy-of-gshhg-vs-srtmgl1/267/15?u=kristofkoch) by @leouieda to come up with an example on how to use OSM coastlines in GMT. Additionally this issue is supposed to serve as an example on how to contribute to the documentation for future reference. As I have never contributed to the documentation, following along this journey might be useful to other novice contributors.

Here is what I got so far. I tried to follow the style of the other gallery examples. Your input on how it can be improved is greatly appreciated. If deemed worthy to become part of the docs I'll have  to figure out on how to actually add it to the repository in a second step.

___

**(XX) OpenStreetMap coastlines in GMT**

Note: This example makes use of data provided by third parties outside the GMT universe. As the internet is constantly changing, the source of this data might move or vanish altogether. 

This example shows how to use coastlines derived from the [OpenStreetMap (OSM) project](https://www.openstreetmap.org/) in GMT. When working with the high resolution SRTM datasets provided by GMT, the GSHHG coastlines might not be precise enough on small scales. Here comes the OSM project to the rescue as it provides high resolution coastlines in the WGS84 datum. Not as convenient and fast as GSHHG but easy enough to give it a try – as long as you have GDAL installed, some space on your hard drive and some processor cycles to spare.

First we need to download the appropriate shape file. The [OSM Wiki](https://wiki.openstreetmap.org/wiki/Shapefiles#Obtaining_shapefiles_from_OSM_data) has some sources, https://osmdata.openstreetmap.de/data/land-polygons.html is very useful. The nice people of the German [FOSSGIS society](https://www.fossgis.de) already did some of the heavylifting and converted the OSM coastlines to a shape file for us. Make sure to get the _Land Polygon_ in the _Large polygons not split_ version with WGS84 datum. It’s a ~630 MB download. 

If your GMT installation comes with GDAL support then you are able to make use of this file right away but GMT converts it under the hood to something more useful on the fly. Neither very fast nor very efficient if you want to use the file multiple times and have to deal with large file sizes as we do here. So we use GDALs `ogr2ogr` and convert the shape file to a native GMT format. 

As the resulting raw ASCII file is very large (>1 GB), GMTs `convert` is used to reduce the file size to a third by converting it to a single-precision binary file. The resulting file is smaller than the initial download while still maintaining precision in the range of centimeters. 

To speed up the plotting process even more, we extract the region of interest with `select`. Finally we are all set to use the coastlines with `plot` to get nice, precise coastlines in our illustrations as presented in figure XX.

```
#!/usr/bin/env bash
# GMT EXAMPLE XX
#
# Purpose:	Illustrate usage of OpenStreetMap coastlines in GMT
# GMT modules:	convert, select, coast, plot
# GDAL progs:	ogr2ogr
#
# First we convert the shapefile to something more GMT friendly. Here we make
# into a large ASCII file. Human-readable but not very efficient. GDALs ogr2ogr
# is used with the "-F GMT" option indicating that the output should have the
# native GMT format. Input is the downloaded land_polygons.shp, and the output
# is named land_polygons_osm_planet.gmt. The resulting file is very large 
# (>1 GB) but already usable. But wait, there is more ...

ogr2ogr -F GMT land_polygons_osm_planet.gmt land_polygons.shp

# Resorting to GMTs convert program we take our very large ASCII file and
# reduce it to about a third of its original size by converting it to a binary
# file.
# Lets have a closer look at -bo2f: 
# -bo selects native binary output
# 2 makes it two data columns
# f indicates 4-byte single-precision for the data

gmt convert land_polygons_osm_planet.gmt -bo2f > land_polygons_osm_planet.bf2

# Again we could use the file as is but there are more performance gains to
# collect. We extract just the area we need for our plot. Here the island of
# Sao Vicente, Cape Verde. select to the rescue! -bo2f is nothing new and -bi2f
# looks very similar: the -bi is indicating native binary input. -R...r gives
# the area which we are interested in.

gmt select land_polygons_osm_planet.bf2 \
 -bi2f -bo2f \
 -R-25.14/16.75/-24.8/16.95r \
 > land_polygons_osm_caboverde.bf2

# Finally it is time to do the thing we are here for: Plot some OSM coastlines
# and compare them to the GSHHG coastlines.

gmt begin caboverde
  
  # To make the lines a bit nicer we set the line endcaps and the way lines are 
  # joined to round
  
  gmt set PS_LINE_CAP round
  gmt set PS_LINE_JOIN round
  
  # To get an idea where the GSHHG coastlines are we lay them down first with a
  # thin red pen. -JL defines a Lambert conic conformal projection and -R...r
  # is the area of the plot. -W defines the pen and -B the style of the plot
  # borders, gridlines and annotations. 
  
  gmt coast -JL-24.9/16.55/16.3/16.7/20c -R-25.14/16.75/-24.8/16.95r \
    -Wred -Ba10mg10m
  
  # Time to use the OSM coastlines we prepared earlier. Straightforward we
  # supply the extracted area, tell plot about the binary format (-bi), define
  # pen (-W) and fill (-G)
  
  gmt plot land_polygons_osm_caboverde.bf2 -bi2f -Wthinnest,black -Ggrey
  
  # Here we plot the GSHHG coastlines a second time but now with a dashed pen
  # to highlight the areas where they would otherwise be hidden behind the grey
  # OSM landmass. 
  
  gmt coast -Wred,dashed
  
gmt end show
```

caboverde

### PR DESCRIPTION
**Pull request by KristofKoch**
**Description of proposed changes**

A new example on how to use coastlines from the OpenStreetMap project in GMT. First pull request ever. Your help is appreciated.

Fixes #3041 

**Reminders**

- [x] Correct base branch selected? `master` for new features, `6.1` for bug fixes
- [x] Make sure that your code follows our style. Use the other functions/files as a basis.

### CODE REVIEW

**Comment by seisman** [comment]
Each example should provide an original PS file in the example directory, i.e. `ex51/ex51.ps`. The PS file will be converted to PNG and shown in the HTML documentation.

**Comment by seisman** [comment]
The biggest problem is that, the data `land_polygons.shp` used in the example is too large and can't be added into the GMT repository. It's OK for users, as they can download the data manually and run the example. 

GMT runs tests nightly to check if the newly generates plots are identical to the original PS files in the GMT repository. Without the data, GMT can't run the test and do the comparison. 

There are two possible solutions:

1. Move the data processing part (i.e., `org2ogr`, `gmtconvert` and `gmtselect`) to the example documentation, and provide the processed smaller file `land_polygons_osm_caboverde.bf2` as a remote cache file.
2. Skip the example in nightly tests.

File: doc/examples/ex51/ex51.sh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  
+  gmt set PS_LINE_CAP round
+  gmt set PS_LINE_JOIN round
+  
+  # To get an idea where the GSHHG coastlines are we lay them down first with a
+  # thin red pen. -JL defines a Lambert conic conformal projection and -R...r
+  # is the area of the plot. -W defines the pen and -B the style of the plot
+  # borders, gridlines and annotations. 
+  
+  gmt coast -JL-24.9/16.55/16.3/16.7/20c -R-25.14/16.75/-24.8/16.95r \
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:57]: Use the new syntax **+r**.
```suggestion
  gmt coast -JL-24.9/16.55/16.3/16.7/20c -R-25.14/16.75/-24.8/16.95+r \
```

**Comment by KristofKoch** [comment]
@seisman thank you for your recommendations. Concerning the nightly tests I'm undecided what to prefer. I see your arguments for both variants. 

However the main part of the example is the preprocessing of the shape file. Providing `land_polygons_osm_caboverde.bf2` would be easy, it is about 12 KB.

I can't make an informed decision here, this is up to more experienced people than I am.

**Comment by seisman** [comment]
@PaulWessel @joa-quim @leouieda Any thoughts? I prefer to skip it in our nightly tests.

**Comment by PaulWessel** [comment]
I think the usefulness of the example lies in the recipe which people would have to apply to their own areas anyway.  So not running this example (at all) seems preferable - just display.

File: doc/rst/source/gallery/ex51.rst [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+If your GMT installation comes with GDAL support then you are able to make use of this file right away but GMT converts it under the hood to something more useful on the fly. Neither very fast nor very efficient if you want to use the file multiple times and have to deal with large file sizes as we do here. So we use GDALs ``ogr2ogr`` and convert the shape file to a native GMT format.
+
+As the resulting raw ASCII file is very large (>1 GB), GMTs :doc:`convert` is used to reduce the file size to a third by converting it to a single-precision binary file. The resulting file is smaller than the initial download while still maintaining precision in the range of centimeters.
+
+To speed up the plotting process even more, we extract the region of interest with :doc:`select`. Finally we are all set to use the coastlines with :doc:`plot` to get nice, precise coastlines in our illustrations.
```
Comment (seisman) [inline doc/rst/source/gallery/ex51.rst:16]: Same reason as above.
```suggestion
To speed up the plotting process even more, we extract the region of interest with :doc:`/gmtselect`. Finally we are all set to use the coastlines with :doc:`/plot` to get nice, precise coastlines in our illustrations.
```

File: doc/examples/ex51/ex51.sh [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+
+# First we convert the shapefile to something more GMT friendly. Here we make
+# into a large ASCII file. Human-readable but not very efficient. GDALs ogr2ogr
+# is used with the "-F GMT" option indicating that the output should have the
+# native GMT format. Input is the downloaded land_polygons.shp, and the output
+# is named land_polygons_osm_planet.gmt. The resulting file is very large 
+# (>1 GB) but already usable. But wait, there is more ...
+
+# ogr2ogr -F GMT land_polygons_osm_planet.gmt land_polygons.shp
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:16]:
```suggestion
ogr2ogr -f OGR_GMT land_polygons_osm_planet.gmt land_polygons.shp
```
1. Uncomment this line is OK, as we'll skip the example in GMT's nightly test
2. Although `-F GMT` may work for you. The ogr2ogr usage tells me to use `-f OGR_GMT` instead. I don't test it. Could you verify if it works for you?

File: doc/examples/ex51/ex51.sh [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+# with higher precision than the GSHHG coastlines.
+# Again, we could use the file as is, but there are more performance gains to
+# collect. We extract just the area we need for our plot. Here the island of
+# Sao Vicente, Cape Verde. select to the rescue! -bo2f is nothing new and -bi2f
+# looks very similar: the -bi is indicating native binary input. -R...r gives
+# the area which we are interested in.
+
+gmt select land_polygons_osm_planet.bf2 \
+ -bi2f -bo2f \
+ -R-25.14/16.75/-24.8/16.95r \
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:38]:
```suggestion
 -R-25.14/16.75/-24.8/16.95+r \
```

File: doc/examples/ex51/ex51.sh [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ > land_polygons_osm_caboverde.bf2
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:39]: This is how the example looks like. 

The code-blocks are designed to be ~120 characters long for displaying long commands Here no need to break a command into short lines.

image

File: doc/examples/ex51/ex51.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+gmt begin ex51
+  
+  # To make the lines a bit nicer we set the line endcaps and the way lines are
+  # joined to round
+  
+  gmt set PS_LINE_CAP round
+  gmt set PS_LINE_JOIN round
+  
+  # To get an idea where the GSHHG coastlines are we lay them down first with a
+  # thin red pen. -JL defines a Lambert conic conformal projection and -R...r
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:53]:
```suggestion
  # thin red pen. -JL defines a Lambert conic conformal projection and -R...+r
```

File: doc/rst/source/gallery/ex51.rst [hunk 7]
```diff
@@ -0,0 +1,25 @@
+.. _example_51:
+
+(51) OpenStreetMap coastlines in GMT
+------------------------------------
```
[the remark this answers is not in the document]
Comment (KristofKoch) [reply doc/rst/source/gallery/ex51.rst:4]: Oh ok, I use soft wrap in my editor to make line length a non-issue. Point taken, will chop up the lines to an acceptable length.

File: doc/examples/ex51/ex51.sh (hunk 3 above)
Comment (KristofKoch) [reply doc/examples/ex51/ex51.sh:16]: I ran both versions and in both cases it works as expected - yours as well as mine:

```
mbp15-kristof:OSM-Kueste kristof$ ogr2ogr -f OGR_GMT land_polygons_osm_planet.gmt land_polygons.shp
Warning 1: The output driver does not seem to natively support Integer64 type for field FID. Converting it to Real instead. -mapFieldType can be used to control field type conversion.
mbp15-kristof:OSM-Kueste kristof$ ogr2ogr -F GMT land_polygons_osm_planet.gmt land_polygons.shp
Warning 1: The output driver does not seem to natively support Integer64 type for field FID. Converting it to Real instead. -mapFieldType can be used to control field type conversion.
mbp15-kristof:OSM-Kueste kristof$
```

the lower-case `-f` is the proper switch but I was unable to find anything about `GMT` vs `OGR_GMT`. 

See
https://gdal.org/programs/ogr2ogr.html#cmdoption-ogr2ogr-f
https://gdal.org/drivers/vector/gmt.html 

So maybe `-f GMT` is the *right* variant?

Comment (KristofKoch) [reply doc/examples/ex51/ex51.sh:16]: Ok, found `OGR_GMT` as valid `format_name` in `ogr2ogr --long-usage`. So `OGR_GMT` it is.

Comment (seisman) [reply doc/examples/ex51/ex51.sh:16]: `ogr2ogr --long-usage` lists `OGR_GMT`, not `GMT`.

`-f GMT` also works for me (GDAL 2.4.4). I'm OK with `-f GMT`.

File: doc/examples/ex51/ex51.sh (hunk 5 above)
Comment (KristofKoch) [reply doc/examples/ex51/ex51.sh:39]: Good point. Aiming for 115 characters line length for long commands. Text still cuts at 80 characters for easier readability.

File: doc/examples/ex51/ex51.sh (hunk 3 above)
Comment (KristofKoch) [reply doc/examples/ex51/ex51.sh:16]: We seem to have replied almost at the same time. I'll take your `OGR_GMT` as the correct version as there is a chance that `GMT` is not recognized and therefore the proper format is guessed from the file extension as per https://gdal.org/programs/ogr2ogr.html#cmdoption-ogr2ogr-f . I think it is better to use documented file formats.

File: doc/examples/ex51/ex51.sh (hunk 6 above)
Comment (KristofKoch) [reply doc/examples/ex51/ex51.sh]: Didn't catch this one …

File: doc/examples/ex51/ex51.sh [hunk 8]
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env bash
+# GMT EXAMPLE 51
+#
+# Purpose:     Illustrate usage of OpenStreetMap coastlines in GMT
+# GMT modules: convert, select, coast, plot
+# GDAL progs:  ogr2ogr
+
+
+# First we convert the shapefile to something more GMT friendly. Here we make it
+# into a large ASCII file. Human-readable but not very efficient. GDALs ogr2ogr
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:10]:
```suggestion
# into a large ASCII file. Human-readable but not very efficient. GDAL's ogr2ogr
```

File: doc/examples/ex51/ex51.sh [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+# is used with the "-f OGR_GMT" option indicating that the output should have
+# the native GMT format. Input is the downloaded land_polygons.shp, and the
+# output is named land_polygons_osm_planet.gmt. The resulting file is very
+# large (>1 GB) but already usable. But wait, there is more ...
+
+ogr2ogr -f OGR_GMT land_polygons_osm_planet.gmt land_polygons.shp
+
+# Resorting to GMTs convert program we take our very large ASCII file and
```
Comment (seisman) [inline doc/examples/ex51/ex51.sh:18]:
```suggestion
# Resorting to GMT's convert program we take our very large ASCII file and
```

File: doc/rst/source/gallery/ex51.rst [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+page <https://osmdata.openstreetmap.de/data/land-polygons.html>`_. It’s a
+~630 MB download.
+
+If your GMT installation comes with GDAL support then you are able to make use
+of this file right away as GMT converts it under the hood to something more
+useful on the fly. Neither very fast nor very efficient if you want to use the
+file multiple times and have to deal with large file sizes as we do here. So we
+use GDALs ``ogr2ogr`` and convert the shape file to a native GMT format.
+
+As the resulting raw ASCII file is very large (>1 GB), GMTs :doc:`/gmtconvert`
```
Comment (seisman) [inline doc/rst/source/gallery/ex51.rst:36]:
```suggestion
As the resulting raw ASCII file is very large (>1 GB), GMT's :doc:`/gmtconvert`
```

File: doc/rst/source/gallery/ex51.rst [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+the OSM coastlines to a shape file for us. Make sure to get the *Land Polygon*
+in the *Large polygons not split* version with WGS84 datum from their `download
+page <https://osmdata.openstreetmap.de/data/land-polygons.html>`_. It’s a
+~630 MB download.
+
+If your GMT installation comes with GDAL support then you are able to make use
+of this file right away as GMT converts it under the hood to something more
+useful on the fly. Neither very fast nor very efficient if you want to use the
+file multiple times and have to deal with large file sizes as we do here. So we
+use GDALs ``ogr2ogr`` and convert the shape file to a native GMT format.
```
Comment (seisman) [inline doc/rst/source/gallery/ex51.rst:34]:
```suggestion
use GDAL's ``ogr2ogr`` and convert the shape file to a native GMT format.
```

File: doc/examples/ex51/ex51.sh (hunk 8 above)
Comment (KristofKoch) [reply doc/examples/ex51/ex51.sh:10]: German Grammar in English text … thanks for catching it.

**Pull request merged**